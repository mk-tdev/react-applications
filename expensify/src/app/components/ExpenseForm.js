import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";
import "react-dates/lib/css/_datepicker.css";

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocued: false,
      errorText: "",
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;

    this.setState(() => ({
      description,
    }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({
      note,
    }));
  };

  onAmountChange = ({ target }) => {
    const amount = target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChanged = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocuesChanged = ({ focused }) => {
    this.setState(() => ({
      calendarFocued: focused,
    }));
  };

  onFormSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description && !this.state.amount) {
      this.setState(() => ({ errorText: "Please provide proper data!" }));
    } else {
      this.setState(() => ({ errorText: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.errorText && (
          <p className="errorText">{this.state.errorText}</p>
        )}
        <form onSubmit={this.onFormSubmit}>
          <div>
            <input
              type="text"
              autoFocus
              placeholder="Description"
              value={this.state.description}
              onChange={this.onDescriptionChange}
            />
          </div>
          <div>
            <input
              type="text"
              placeholder="Amount"
              value={this.state.amount}
              onChange={this.onAmountChange}
            />
          </div>
          <div>
            <textarea
              placeholder="Add a note about it"
              value={this.state.note}
              onChange={this.onNoteChange}
            ></textarea>
          </div>

          <div>
            <SingleDatePicker
              date={this.state.createdAt}
              onDateChange={this.onDateChanged}
              focused={this.state.calendarFocued}
              onFocusChange={this.onFocuesChanged}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>

          <div>
            <button>
              {this.props.expense ? "Edit Expense" : "Add Expense"}
            </button>
          </div>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
