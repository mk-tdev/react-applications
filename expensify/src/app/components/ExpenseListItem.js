import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import moment from "moment";

import { startRemoveExpenses } from "../actions/expenses";

export const ExpenseListItem = ({
  description,
  amount,
  createdAt,
  id,
  dispatch,
}) => {
  // console.log("moment", moment(createdAt).date());

  return (
    <div className="expense-item">
      <div>
        <Link to={`/edit/${id}`}>
          <h4>{description}</h4>
        </Link>
        <p>
          ${amount} - (at: {createdAt})
        </p>
      </div>
      <button
        onClick={() => {
          dispatch(startRemoveExpenses({ id }));
        }}
      >
        Remove
      </button>
    </div>
  );
};

export default connect()(ExpenseListItem);
