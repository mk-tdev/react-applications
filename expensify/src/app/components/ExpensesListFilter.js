import React from "react";
import { connect } from "react-redux";

import {
  setEndDate,
  setStartDate,
  setTextFilter,
  sortByAmount,
  sortByDate,
} from "../actions/filters";

class ExpensesListFilter extends React.Component {
  state = {
    calendarFocued: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };

  onFocuesChanged = ({ focused }) => {
    this.setState(() => ({
      calendarFocued: focused,
    }));
  };

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={(e) => {
            this.props.dispatch(setTextFilter(e.target.value));
          }}
        />

        <select
          value={this.props.filters.sortBy}
          onChange={(e) => {
            if (e.target.value === "date") {
              this.props.dispatch(sortByDate());
            } else {
              this.props.dispatch(sortByAmount());
            }
          }}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>

        {/* <DateRangePicker
          startDateId="start_date_input"
          endDateId="end_date_input"
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          onFocusChange={this.onFocuesChanged}
          focusedInput={this.state.calendarFocued}
          isOutsideRange={() => false}
        /> */}
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

export default connect(mapStateToProps)(ExpensesListFilter);
