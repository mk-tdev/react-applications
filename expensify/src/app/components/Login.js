import React from "react";
import { connect } from "react-redux";
import { startLogin } from "../actions/auth";

export const Login = ({ startLogin }) => {
  return (
    <div>
      <p>Login</p>

      <button onClick={startLogin}>Login</button>
    </div>
  );
};

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin()),
});

export default connect(undefined, mapDispatchToProps)(Login);
