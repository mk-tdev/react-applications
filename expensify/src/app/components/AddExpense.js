import React from "react";
import { connect } from "react-redux";
import { useHistory } from "react-router-dom";

import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../actions/expenses";

function AddExpense(props) {
  const history = useHistory();
  return (
    <div>
      <h1>Add Expense</h1>
      <ExpenseForm
        onSubmit={(expense) => {
          props.dispatch(addExpense(expense));
          history.push("/");
        }}
      />
    </div>
  );
}

export default connect()(AddExpense);
