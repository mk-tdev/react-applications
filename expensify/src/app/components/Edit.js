import React from "react";
import { connect } from "react-redux";
import { useHistory } from "react-router-dom";
import ExpenseForm from "./ExpenseForm";
import { editExpense } from "../actions/expenses";

const Edit = (props) => {
  const history = useHistory();

  return (
    <div>
      <h2>Edit</h2>

      <ExpenseForm
        expense={props.expense}
        onSubmit={(expense) => {
          console.log("expense: ", expense);
          props.dispatch(editExpense(props.expense.id, expense));
          history.push("/");
        }}
      />
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((exp) => exp.id === props.match.params.id),
  };
};

export default connect(mapStateToProps)(Edit);
