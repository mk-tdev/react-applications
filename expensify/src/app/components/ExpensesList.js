import React from "react";

import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

function ExpensesList(props) {
  return (
    <div>
      {props.expenses.map((exp) => {
        return <ExpenseListItem key={exp.id} {...exp} />;
      })}
    </div>
  );
}

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpensesList);
