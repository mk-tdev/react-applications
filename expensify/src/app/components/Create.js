import React from "react";
import AddExpense from "./AddExpense";

function Create() {
  return <AddExpense />;
}

export default Create;
