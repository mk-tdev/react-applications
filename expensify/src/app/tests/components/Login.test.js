import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";
configure({ adapter: new Adapter() });

import { Login } from "../../components/Login";

test("should render NotFound", () => {
  const wrapper = shallow(<Login />);
  expect(wrapper).toMatchSnapshot();
});
