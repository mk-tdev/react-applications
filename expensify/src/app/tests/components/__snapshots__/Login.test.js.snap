// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render NotFound 1`] = `
<div>
  <p>
    Login
  </p>
  <button>
    Login
  </button>
</div>
`;
